src/globus_sdk/services/auth/consents.py

```diff
@@ -0,0 +1,311 @@
+        """
+        Populate the nodes and edges of the tree by traversing the forest.
+
+        Nodes/Edges are included in the tree iff they are reachable from the root.
+        """
+        nodes_to_evaluate = {self.root.id}
+        while nodes_to_evaluate:
+            consent_id = nodes_to_evaluate.pop()
+            consent = forest.get_node(consent_id)
+            self.edges[consent.id] = forest.edges[consent.id]
+
+            for child_id in self.edges[consent.id]:
+                if child_id not in self._node_by_id:
+                    self.nodes.append(forest.get_node(child_id))
+                    self._node_by_id[child_id] = forest.get_node(child_id)
+                    nodes_to_evaluate.add(child_id)
+
+    def get_node(self, consent_id: int) -> Consent:
+        return self._node_by_id[consent_id]
+
+    def meets_scope_requirements(self, scope: Scope) -> bool:
+        """
+        Check whether this consent tree meets a particular scope requirement.
+
+        :param scope: A Scope requirement to check against the tree.
+        """
+        return self._meets_scope_requirements_recursive(self.root, scope)
+
+    def _meets_scope_requirements_recursive(self, node: Consent, scope: Scope) -> bool:
+        """
+        Check recursively whether a consent node meets the scope requirements defined
+            by a scope object.
+        """
+        if node.scope_name != scope.scope_string:
+            return False
+
+        for dependent_scope in scope.dependencies:
+            for child_id in self.edges[node.id]:
+                if self._meets_scope_requirements_recursive(
+                    self.get_node(child_id), dependent_scope
+                ):
+                    # We found a child containing this full dependent scope tree
+                    # Move onto the next dependent scope tree
+                    break
+            else:
+                # We didn't find any child containing this full dependent scope tree
+                return False
+        # We found at least one child containing each full dependent scope tree
+        return True
+
+    @property
+    def max_depth(self) -> int:
+        return self._max_depth_recursive(self.root, 1)
+
+    def _max_depth_recursive(self, node: Consent, depth: int) -> int:
+        if len(self.edges[node.id]) == 0:
+            return depth
+        return max(
+            self._max_depth_recursive(self.get_node(child_id), depth + 1)
+            for child_id in self.edges[node.id]
+        )
+
+    def pprint(self) -> None:
+        """Pretty print a textual representation of the tree (one line per node)."""
+        self._pprint_recursive(self.root, 0)
+
+    def _pprint_recursive(self, node: Consent, tab_depth: int) -> None:
+        print(f"{' ' * tab_depth} - {node}")
+        for child_id in self.edges[node.id]:
+            self._pprint_recursive(self.get_node(child_id), tab_depth + 2)
```

Comment:
I put it into a distinct path because it was so long.

A typical call here will print out 1 - 8 new lines (all of which are very wide). I don't have perfect heuristic for "when is a string too big for __str__" but this seemed like it was pushing that line.

Reply:
Moved to `__str__`